File: VB/GrowthWare.WebSupport/Utilities/SecurityEntityUtility.vb
Imports GrowthWare.Framework.Model.Profiles
Imports GrowthWare.Framework.Common
Imports System.Collections.ObjectModel
Imports System.Globalization
Imports System.Web
Imports GrowthWare.Framework.BusinessData.BusinessLogicLayer
Imports GrowthWare.Framework.BusinessData.DataAccessLayer

Namespace Utilities
    ''' <summary>
    ''' SecurityEntityUtility serves as the focal point for any web application needing to utiltize the GrowthWare framework.
    ''' Web needs such as caching are handeled here
    ''' </summary>
    Public Module SecurityEntityUtility
        Private s_DefaultProfile As MSecurityEntityProfile = Nothing
        'Private Shared m_BSecurityEntity As BSecurityEntity = Nothing
        Private s_CacheName As String = "SecurityEntityProfiles"

        ''' <summary>
        ''' Creates and returns MSecurityEntityProfile populated with information from the
        ''' configuration file.
        ''' </summary>
        ''' <returns>MSecurityEntityProfile</returns>
        Public Function DefaultProfile() As MSecurityEntityProfile
            If s_DefaultProfile Is Nothing Then
                Dim mDefaultProfile As MSecurityEntityProfile = New MSecurityEntityProfile()
                mDefaultProfile.Id = Integer.Parse(ConfigSettings.DefaultSecurityEntityId.ToString(), CultureInfo.InvariantCulture)
                mDefaultProfile.DataAccessLayer = ConfigSettings.DataAccessLayer
                mDefaultProfile.DataAccessLayerNamespace = ConfigSettings.DataAccessLayerNamespace(mDefaultProfile.DataAccessLayer)
                mDefaultProfile.DataAccessLayerAssemblyName = ConfigSettings.DataAccessLayerAssemblyName(mDefaultProfile.DataAccessLayer)
                mDefaultProfile.ConnectionString = ConfigSettings.ConnectionString(mDefaultProfile.DataAccessLayer)
                s_DefaultProfile = mDefaultProfile
            End If
            Return s_DefaultProfile
        End Function

        ''' <summary>
        ''' Returns the current MSecurityEntityProfile from context.  If one is not found in context then 
        ''' the default values from the config file will be returned.
        ''' </summary>
        ''' <returns>MSecurityEntityProfile</returns>
        Public Function CurrentProfile() As MSecurityEntityProfile
            Dim mAccount As String = AccountUtility.HttpContextUserName
            Dim mClientChoicesState As MClientChoicesState = ClientChoicesUtility.GetClientChoicesState(mAccount)
            Dim mRetVal As MSecurityEntityProfile = Nothing
            If mClientChoicesState IsNot Nothing Then
                Dim mCurrentSecurityEntityID As Integer = Integer.Parse(mClientChoicesState(MClientChoices.SecurityEntityId).ToString(), CultureInfo.InvariantCulture)
                Dim mProfiles As Collection(Of MSecurityEntityProfile)
                mProfiles = Profiles()

                Dim mResult = From mProfile In mProfiles Where mProfile.Id = mCurrentSecurityEntityID Select mProfile
                mRetVal = mResult.First
            End If
            If mRetVal Is Nothing Then mRetVal = DefaultProfile()
            Return mRetVal
        End Function

        ''' <summary>
        ''' Gets the profiles.
        ''' </summary>
        ''' <returns>Collection{MSecurityEntityProfile}.</returns>
        Public Function Profiles() As Collection(Of MSecurityEntityProfile)
            Dim mSecurityEntityCollection As Collection(Of MSecurityEntityProfile) = Nothing
            mSecurityEntityCollection = CType(HttpContext.Current.Cache(s_CacheName), Collection(Of MSecurityEntityProfile))
            If mSecurityEntityCollection Is Nothing Then
                Dim mBSecurityEntity As BSecurityEntity = New BSecurityEntity(DefaultProfile(), ConfigSettings.CentralManagement)
                mSecurityEntityCollection = mBSecurityEntity.SecurityEntities()
                CacheController.AddToCacheDependency(s_CacheName, mSecurityEntityCollection)
            End If
            Return mSecurityEntityCollection
        End Function

        ''' <summary>
        ''' Gets the profile.
        ''' </summary>
        ''' <param name="securityEntityId">The security entity ID.</param>
        ''' <returns>MAccountProfile.</returns>
        Public Function GetProfile(ByVal securityEntityId As Integer) As MSecurityEntityProfile
            Dim mResult = From mProfile In Profiles() Where mProfile.Id = securityEntityId Select mProfile
            Dim mRetVal As MSecurityEntityProfile = New MSecurityEntityProfile()
            Try
                mRetVal = mResult.First
            Catch ex As NullReferenceException
                mRetVal = Nothing
            End Try
            Return mRetVal
        End Function

        ''' <summary>
        ''' Gets the profile.
        ''' </summary>
        ''' <param name="name">The name.</param>
        ''' <returns>MSecurityEntityProfile.</returns>
        Public Function GetProfile(ByVal name As String) As MSecurityEntityProfile
            Dim mResult = From mProfile In Profiles() Where mProfile.Name.ToLower(CultureInfo.CurrentCulture) = name.ToLower(CultureInfo.CurrentCulture) Select mProfile
            Dim mRetVal As MSecurityEntityProfile = New MSecurityEntityProfile()
            Try
                mRetVal = mResult.First
            Catch ex As NullReferenceException
                mRetVal = Nothing
            End Try
            Return mRetVal
        End Function

        ''' <summary>
        ''' Gets the valid security entities.
        ''' </summary>
        ''' <param name="account">The account.</param>
        ''' <param name="securityEntityId">The security entity ID.</param>
        ''' <param name="IsSecurityEntityAdministrator">if set to <c>true</c> [is security entity administrator].</param>
        ''' <returns>DataView.</returns>
        Public Function GetValidSecurityEntities(ByVal account As String, ByVal securityEntityId As Integer, ByVal isSecurityEntityAdministrator As Boolean) As DataView
            Dim mBSecurityEntity As BSecurityEntity = New BSecurityEntity(CurrentProfile(), ConfigSettings.CentralManagement)
            Return mBSecurityEntity.GetValidSecurityEntities(account, securityEntityId, isSecurityEntityAdministrator).DefaultView()
        End Function

        ''' <summary>
        ''' Saves the specified profile.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        Public Sub Save(ByVal profile As MSecurityEntityProfile)
            Try
                Dim mBSecurityEntity As BSecurityEntity = New BSecurityEntity(CurrentProfile(), ConfigSettings.CentralManagement)
                mBSecurityEntity.Save(profile)
                CacheController.RemoveAllCache()
            Catch ex As DataAccessLayerException
                Dim mLog As Logger = Logger.Instance()
                mLog.Error(ex)
            End Try
        End Sub

        ''' <summary>
        ''' Returns a datatable of the search data
        ''' </summary>
        ''' <param name="searchCriteria">MSearchCriteria</param>
        ''' <returns>NULL/Nothing if no records are returned.</returns>
        ''' <remarks></remarks>
        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable
            Try
                Dim mBSecurityEntity As BSecurityEntity = New BSecurityEntity(CurrentProfile(), ConfigSettings.CentralManagement)
                Return mBSecurityEntity.Search(searchCriteria)
            Catch ex As IndexOutOfRangeException
                'no data is not a problem
                Return Nothing
            End Try
        End Function

    End Module
End Namespace

File: Net_Framework/CS/GrowthWare.WebSupport/Utilities/SecurityEntityUtility.cs
using GrowthWare.Framework.BusinessData.BusinessLogicLayer;
using GrowthWare.Framework.BusinessData.DataAccessLayer;
using GrowthWare.Framework.Common;
using GrowthWare.Framework.Model.Profiles;
using System;
using System.Collections.ObjectModel;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Web;

namespace GrowthWare.WebSupport.Utilities
{
    /// <summary>
    /// SecurityEntityUtility serves as the focal point for any web application needing to utiltize the GrowthWare framework.
    /// Web needs such as caching are handeled here
    /// </summary>
    public static class SecurityEntityUtility
    {
        private static String s_CacheName = "Cached_SecurityEntities";

        private static MSecurityEntityProfile s_DefaultProfile = null;

        //private static BSecurityEntity m_BSecurityEntity = null;

        private const string SESSION_PROFILE_NAME = "SessionSecurityEntityProfile";

        /// <summary>
        /// Creates and returns MSecurityEntityProfile populated with information from the
        /// configuration file.
        /// </summary>
        /// <returns>MSecurityEntityProfile</returns>
        public static MSecurityEntityProfile DefaultProfile()
        {
            if (s_DefaultProfile == null)
            {
                MSecurityEntityProfile mDefaultProfile = new MSecurityEntityProfile();
                mDefaultProfile.Id = int.Parse(ConfigSettings.DefaultSecurityEntityId.ToString(CultureInfo.InvariantCulture), CultureInfo.InvariantCulture);
                mDefaultProfile.DataAccessLayer = ConfigSettings.DataAccessLayer;
                mDefaultProfile.DataAccessLayerNamespace = ConfigSettings.DataAccessLayerNamespace(mDefaultProfile.DataAccessLayer);
                mDefaultProfile.DataAccessLayerAssemblyName = ConfigSettings.DataAccessLayerAssemblyName(mDefaultProfile.DataAccessLayer);
                mDefaultProfile.ConnectionString = ConfigSettings.ConnectionString(mDefaultProfile.DataAccessLayer);
                s_DefaultProfile = mDefaultProfile;
            }
            return s_DefaultProfile;
        }

        /// <summary>
        /// Returns the current MSecurityEntityProfile from context.  If one is not found in context then 
        /// the default values from the config file will be returned.
        /// </summary>
        /// <returns>MSecurityEntityProfile</returns>
        public static MSecurityEntityProfile CurrentProfile()
        {
            MSecurityEntityProfile mRetProfile = null;
            String mAccount = AccountUtility.HttpContextUserName();
            MClientChoicesState mClientChoicesState = ClientChoicesUtility.GetClientChoicesState(mAccount);
            if (mClientChoicesState != null) 
            {
                int mSecurityEntity = int.Parse(mClientChoicesState[MClientChoices.SecurityEntityId].ToString(), CultureInfo.InvariantCulture);
                mRetProfile = GetProfile(mSecurityEntity);
            }
            if (mRetProfile == null) mRetProfile = DefaultProfile();
            return mRetProfile;
        }

        /// <summary>
        /// Retrieves all security entities from the either the database or cache
        /// </summary>
        /// <returns>A Collection of MSecurityEntityProfile</returns>
        public static Collection<MSecurityEntityProfile> Profiles()
        {
            Collection<MSecurityEntityProfile> mRetVal = null;
            mRetVal = (Collection<MSecurityEntityProfile>)(HttpContext.Current.Cache[s_CacheName]);
            if (mRetVal == null)
            {
                BSecurityEntity mBSecurityEntity = new BSecurityEntity(DefaultProfile(), ConfigSettings.CentralManagement);
                mRetVal = mBSecurityEntity.SecurityEntities();
                CacheController.AddToCacheDependency(s_CacheName, mRetVal);
            }
            return mRetVal;
        }

        /// <summary>
        /// Get a single function given it's action.
        /// </summary>
        /// <param name="name">String</param>
        /// <returns>MSecurityEntityProfile</returns>
        public static MSecurityEntityProfile GetProfile(String name)
        {
            MSecurityEntityProfile mRetVal = new MSecurityEntityProfile();
            var mResult = from mProfile in Profiles()
                          where mProfile.Name.ToLower(CultureInfo.CurrentCulture) == name.ToLower(CultureInfo.CurrentCulture)
                          select mProfile;
            try
            {
                mRetVal = mResult.First();
            }
            catch (NullReferenceException)
            {
                mRetVal = null;
            }
            return mRetVal;

        }

        /// <summary>
        /// Get a single function given it's id.
        /// </summary>
        /// <param name="securityEntitySeqId">int or Integer</param>
        /// <returns>MSecurityEntityProfile</returns>
        public static MSecurityEntityProfile GetProfile(int securityEntitySeqId)
        {
            MSecurityEntityProfile mRetVal = new MSecurityEntityProfile();
            var mResult = from mProfile in Profiles()
                          where mProfile.Id == securityEntitySeqId
                          select mProfile;
            try
            {
                mRetVal = mResult.First();
            }
            catch (NullReferenceException)
            {
                mRetVal = null;
            }
            return mRetVal;
        }

        /// <summary>
        /// Gets the valid security entities.
        /// </summary>
        /// <param name="account">The account.</param>
        /// <param name="securityEntityId">The security entity id.</param>
        /// <param name="isSystemAdmin">if set to <c>true</c> [is system admin].</param>
        /// <returns>DataView.</returns>
        public static DataView GetValidSecurityEntities(string account, int securityEntityId, bool isSystemAdmin)
        {
            BSecurityEntity mBSecurityEntities = new BSecurityEntity(SecurityEntityUtility.CurrentProfile(), ConfigSettings.CentralManagement);
            return mBSecurityEntities.GetValidSecurityEntities(account, securityEntityId, isSystemAdmin).DefaultView;
        }

        /// <summary>
        /// Searches Security Entities with the specified search criteria.
        /// </summary>
        /// <param name="searchCriteria">The search criteria.</param>
        /// <returns>DataTable.</returns>
        public static DataTable Search(MSearchCriteria searchCriteria)
        {
            try
            {
                BSecurityEntity mBSecurityEntities = new BSecurityEntity(SecurityEntityUtility.CurrentProfile(), ConfigSettings.CentralManagement);
                return mBSecurityEntities.Search(searchCriteria);
            }
            catch (IndexOutOfRangeException)
            {
                return null;
            }
        }

        /// <summary>
        /// Saves the specified profile.
        /// </summary>
        /// <param name="profile">The profile.</param>
        public static void Save(MSecurityEntityProfile profile)
        {
            try
            {
                BSecurityEntity mBSecurityEntities = new BSecurityEntity(SecurityEntityUtility.CurrentProfile(), ConfigSettings.CentralManagement);
                mBSecurityEntities.Save(profile);
                CacheController.RemoveAllCache();
            }
            catch (DataAccessLayerException ex)
            {
                Logger mLog = Logger.Instance();
                mLog.Error(ex);
            }
        }
    }
}
